albatross/core/model.h

```diff
@@ -13,266 +13,110 @@
-  }
-
-  /*
-   * Provides a wrapper around the implementation `fit_` which performs
-   * simple size checks and makes sure the fit method is called before
-   * predict.
-   */
-  void fit(const std::vector<FeatureType> &features,
-           const MarginalDistribution &targets) {
-    assert(features.size() > 0);
-    assert(features.size() == static_cast<std::size_t>(targets.size()));
-    has_been_fit_ = true;
-    insights_["input_feature_count"] = std::to_string(features.size());
-    fit_(features, targets);
-  }
-
-  /*
-   * Convenience function which assumes zero target covariance.
-   */
-  void fit(const std::vector<FeatureType> &features,
-           const Eigen::VectorXd &targets) {
-    return fit(features, MarginalDistribution(targets));
-  }
-
-  /*
-   * Convenience function which unpacks a dataset into features and targets.
-   */
-  void fit(const RegressionDataset<FeatureType> &dataset) {
-    return fit(dataset.features, dataset.targets);
-  }
-
-  /*
-   * Similar to fit, this predict methods wrap the implementation `predict_*_`
-   * and makes simple checks to confirm the implementation is returning
-   * properly sized Distribution.
-   */
-  template <typename PredictType = JointDistribution>
-  PredictType predict(const std::vector<FeatureType> &features) const {
-    return predict(features, detail::PredictTypeIdentity<PredictType>());
-  }
-
-  template <typename PredictType = JointDistribution>
-  PredictType predict(const FeatureType &feature) const {
-    std::vector<FeatureType> features = {feature};
-    return predict<PredictType>(features);
-  }
-
-  template <typename PredictType = MarginalDistribution>
-  std::vector<PredictType>
-  cross_validated_predictions(const RegressionDataset<FeatureType> &dataset,
-                              const FoldIndexer &fold_indexer) {
-    return cross_validated_predictions_(
-        dataset, fold_indexer, detail::PredictTypeIdentity<PredictType>());
-  }
-
-  // Because cross validation can never properly produce a full
-  // joint distribution it is common to only use the marginal
-  // predictions, hence the different default from predict.
-  template <typename PredictType = MarginalDistribution>
-  std::vector<PredictType> cross_validated_predictions(
-      const std::vector<RegressionFold<FeatureType>> &folds) {
-    // Iteratively make predictions and assemble the output vector
-    std::vector<PredictType> predictions;
-    for (std::size_t i = 0; i < folds.size(); i++) {
-      fit(folds[i].train_dataset);
-      predictions.push_back(
-          predict<PredictType>(folds[i].test_dataset.features));
-    }
-    return predictions;
-  }
+template <typename ModelType> class ModelBase : public ParameterHandlingMixin {
 
-  std::string pretty_string() const {
-    std::ostringstream ss;
-    ss << get_name() << std::endl;
-    ss << ParameterHandlingMixin::pretty_string();
-    return ss.str();
-  }
+  template <typename X, typename Y, typename Z>
```

Comment:
These friend classes are required in order to allow things like `Prediction` to inspect a `ModelType` and decide which `predict_` methods are available.  This way we can keep the CRTP created `fit_` and `predict_` methods private to avoid confusion.